File: inorder_cartesian_tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 int max_idx(vector<int> &A, int start, int end){
     int max = A[start];
     int idx = start;
     for(int i = start+1; i<= end; i++){
         if (A[i]> max){
             max = A[i];
             idx = i;
         } 
     }
     return idx;
 }
 TreeNode* cart_tree(vector<int> &A, int start, int end){
     if(start>end) return NULL;
     int max = max_idx(A, start, end);
     TreeNode* x = new TreeNode(A[max]);
     x->right = cart_tree(A,max+1,end);
     x->left = cart_tree(A,start,max-1);
     return x;
 }
 TreeNode* Solution::buildTree(vector<int> &A) {
    // Do not write main() function.
    // Do not read input, instead use the arguments to the function.
    // Do not print the output, instead return values as specified
    // Still have a doubt. Checkout www.interviewbit.com/pages/sample_codes/ for more details
     return cart_tree(A,0,A.size()-1);
    
}

File: inorder_cartesian_tree.py
#Given an inorder traversal of a cartesian tree, construct the tree.
#
# Cartesian tree : is a heap ordered binary tree, where the root is greater than all the elements in the subtree. 
# Note: You may assume that duplicates do not exist in the tree. 
#Example :
#
#Input : [1 2 3]
#
#Return :   
#          3
#         /
#        2
#       /
#      1
#
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param A : list of integers
    # @return the root node in the tree
    def buildTree(self, A):
        def create_tree(node,idx,A):
            if idx > 0:
                left_val = max(A[:idx])
                left_idx = A[:idx].index(left_val)
                left_node = TreeNode(left_val)
                node.left = left_node
                create_tree(left_node,left_idx,A[:idx])
            if idx < len(A) - 1:
                right_val = max(A[idx+1:])
                right_idx = A[idx+1:].index(right_val)
                right_node = TreeNode(right_val)
                node.right = right_node
                create_tree(right_node,right_idx,A[idx+1:])
            return
        if len(A) == 0:
            return TreeNode(0)
        my_root_val = max(A)
        root_idx = A.index(my_root_val)
        my_root = TreeNode(my_root_val)
        create_tree(my_root,root_idx,A)
        return my_root
